### PR TITLE
add DataRangeWidget to define DataGrid to show

### PR DESCRIPTION
**Pull request by AlexTcots**
add DataRangeWidget to define row_begin,row_end, columns_begin and columns_end for DataGrid

### CODE REVIEW

File: src/main.ipynb [hunk 1]
```diff
@@ -52,9 +53,23 @@
...
+       "version_major": 2,
+       "version_minor": 0,
+       "model_id": "2d8daa9dfd014861b4b7216049c13d3e"
+      }
+     },
+     "execution_count": 7,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
```
Comment (quantum-entangled) [inline src/main.ipynb:72]: `*.ipynb` metadata changes are "junk" changes for VCS, I'm working on automatic pre-comit tool for clearing all outputs and execution counts, but for now you should do it manually and not add them to commits.

**Review by quantum-entangled — CHANGES_REQUESTED** [review]
Sorry for taking to long, I've reviewed your PR and commented several things. In general, this functionality shouldn't be implemented separately from existing DataGrid widget, so please remake this a little bit and also note several app architecture changes from recent commits.

File: src/Managers/DataManager.py [hunk 2]
```diff
@@ -16,17 +16,32 @@ def __init__(self, data: Data, model: Model) -> None:
...
+        self._range[0] = begin_r
+        self._range[1] = end_r
+        self._range[2] = begin_c
+        self._range[3] = end_c
+        with output_handler:
+            out_txt="Rows from {0} to {1} and Columns from {2} to {3}"
+            print(out_txt.format(self._range[0],
+                                 self._range[1],
+                                 self._range[2],
+                                 self._range[3]))
```
[the remark this answers is not in the document]
Comment (AlexTcots) [reply src/Managers/DataManager.py:40]: OK, I will remake it.

**Comment by AlexTcots** [comment]
yes, please review the new commits

**Comment by quantum-entangled** [comment]
Ok, I will do it soon. If you'll be ready for a review next time you change some code, please, assign me as a reviewer first time or re-request review in this section, cause otherwise I won't receive any notifications.
image

**Comment by AlexTcots** [comment]
Ok, I got it.

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows

File: src/Managers/DataManager.py [hunk 3]
```diff
@@ -16,17 +16,32 @@ def __init__(self, data: Data, model: Model) -> None:
...
+
 
     def upload_file(self, file_path: Any) -> None:
         """Read file to the pandas format and store it."""
         self._data.file = pd.read_csv(
             filepath_or_buffer=file_path, header=0, skipinitialspace=True
         )
         self._data.columns = list(self._data.file.columns)
+    def set_range_grid(self, begin_r: Protocol, end_r: Protocol,
+                       begin_c: Protocol, end_c: Protocol) -> None:
```
Comment (quantum-entangled) [inline src/Managers/DataManager.py:30]: I'm not sure why do you use `Protocol`s here? I assume these values are just `int`s.

File: src/UI/CustomWidgets/DataPreparation/DataGridWidget.py [hunk 4]
```diff
@@ -48,3 +86,8 @@ def _on_show_grid_button_clicked(self, _) -> None:
     def _on_widget_state_change(self) -> None:
         """Callback for parent widget ensemble."""
         self.grid_output.clear_output()
+
+    def _on_set_grid_button_clicked(self, _) -> None:
+        self.data_manager.set_range_grid(self._row_range.value[0], self._row_range.value[1],
+                                         self._col_range.value[0], self._col_range.value[1]
+                                         )
```
Comment (quantum-entangled) [inline src/UI/CustomWidgets/DataPreparation/DataGridWidget.py:93]: Using ranges of columns and rows as parameters for `show_data_grid()`, you won't need this function at all.

File: src/Managers/DataManager.py [hunk 5]
```diff
@@ -16,17 +16,32 @@ def __init__(self, data: Data, model: Model) -> None:
...
+        self._range[0] = begin_r
+        self._range[1] = end_r
+        self._range[2] = begin_c
+        self._range[3] = end_c
 
+        out_txt="Rows from {0} to {1} and Columns from {2} to {3}"
+        print(out_txt.format(self._range[0],
+                                 self._range[1],
+                                 self._range[2],
+                                 self._range[3]))
```
Comment (quantum-entangled) [inline src/Managers/DataManager.py:40]: I think we don't really need to print ranges here, because they will be shown on sliders anyway.

**Review by quantum-entangled — CHANGES_REQUESTED** [review]
I reviewed your changes, some things should be simplified still. Please, check them out and ask any questions, if you have some.

**Comment by quantum-entangled** [comment]
Good job, closing it.

**Pull request merged**